### PR TITLE
feat: bucket migration state improvement

### PR DESCRIPTION
**Pull request by jingjunLi**
### Description

Optimizations for bucket migration:
  * Persistence of Quota State
  * Optimization of state persistence in DB for each phase of bucket migration
  * Bucket Migration Progress Query
  * Confirming Complete Event & RejectMigrateBucketEvent
  * Bucket Migration GC Trigger

### Feat 1:  **Persistence of Quota State**

In the previous setup, the source storage provider (Src SP) utilized an in-memory map (`ManageModular::migratingBuckets`) to keep track of whether the quota for bucket migration had been deducted. However, this approach presented challenges in scenarios where the source storage provider experienced an outage and needed to restart, making it difficult to maintain accurate state. To address this issue, the in-memory map has been replaced with a persistent MySQL table.
image

Abstract some logic from **preMigrateBucketHandler** and **postMigrateBucketHandler** into **Manager::NotifyPreMigrateBucketAndDeductQuota & Manager::NotifyPostMigrateBucketAndRecoupQuota**.

- 1) After Src SP deducts Quota, the corresponding state is persisted in the database, and a check is performed. This mainly involves the deducted quota amount and the recouped quota amount.
- 2) Src SP recoups Quota.
image

### feat 2: Optimization of state persistence in DB for each phase of bucket migration

- Introduction of the **`MigrateBucketProgressTable`** table to record the state and meta of bucket migration. Additionally, encapsulation of interfaces for modifying state and meta.

### feat 3: Bucket Migration Progress Query

In both **`MigrateGVGUnitMeta`** (core/spdb/entity.go) and **`MigrateGVGTable`**, add a field named **`MigratedBytesSize`**.

1. Upon each successful migration, accumulate the **`MigratedBytesSize`** from the task. When reporting and calling **`updateMigrateGVGStatus`** during the update, use **`UpdateMigrateGVGMigratedBytesSize`** to update the **`MigratedBytesSize`** field.
2. Users can query the progress of all bucket migrations through **`listExecutePlan`**.

### Example

NA

### Changes

Notable changes: 

**Notification from Destination SP to Source SP - PreMigrateBucket:**
In the processing flow (preMigrateBucketHandler) at the source SP upon receiving notification:

1) Query the `MigrateBucketTable` for the `MigrateStatus` status. If the quota has already been deducted, return "ok" directly.
2) If quota deduction is pending, invoke `DeductQuotaForBucketMigrate` to deduct the quota.
3) After quota deduction, call `NotifyDonePreMigrateBucket` to update the `MigrateStatus` status in the `MigrateBucketTable` to "PreNotified."

**Notification from Destination SP to Source SP - PostMigrateBucket:**
- `PostMigrateBucket` can only be invoked once.

### Potential Impacts
* add potential impacts for other components here
* ...

### CODE REVIEW

File: base/gfspclient/interface.go [hunk 1]
```diff
@@ -88,8 +88,9 @@ type ManagerAPI interface {
 	PickVirtualGroupFamilyID(ctx context.Context, task coretask.ApprovalCreateBucketTask) (uint32, error)
 	NotifyMigrateSwapOut(ctx context.Context, swapOut *virtualgrouptypes.MsgSwapOut) error
 	GetTasksStats(ctx context.Context) (*gfspserver.TasksStats, error)
-	NotifyPreMigrateBucket(ctx context.Context, bucketID uint64) error
-	NotifyPostMigrateBucket(ctx context.Context, bmInfo *gfsptask.GfSpBucketMigrationInfo) error
+	GetMigrationBucketState(ctx context.Context, bucketID uint64) (*gfspserver.MigrateBucketProgressMeta, error)
+	NotifyPostMigrateBucketAndRecoupQuota(ctx context.Context, bmInfo *gfsptask.GfSpBucketMigrationInfo) (*gfsptask.GfSpBucketQuotaInfo, error)
+	NotifyPreMigrateBucketAndDeductQuota(ctx context.Context, bucketID uint64) (*gfsptask.GfSpBucketQuotaInfo, error)
```
Comment (sysvm) [inline base/gfspclient/interface.go:93]: Which word is better: Migrate or Migration?

File: modular/manager/migrate_unit.go [hunk 2]
```diff
@@ -8,14 +8,39 @@ import (
 )
 
 type MigrateStatus int32
+type BucketMigrateState int32
```
Comment (sysvm) [inline modular/manager/migrate_unit.go:11]:
```go
type (
	MigrateStatus      int32
	BucketMigrateState int32
)
```

File: proto/base/types/gfsptask/task.proto [hunk 3]
```diff
@@ -201,6 +201,12 @@ message GfSpMigratePieceTask {
 message GfSpGCBucketMigrationTask {
   GfSpTask task = 1;
   uint64 bucket_id = 2;
+  uint64 last_gc_object_id = 3;
+  uint64 last_gc_gvg_id = 4;
+  // whether the gc of bucket migration succeed
+  bool finished = 5;
+  uint64 gvg_num_total = 6; // the bucket migration gc total gvg num
```
Comment (sysvm) [inline proto/base/types/gfsptask/task.proto:208]: maybe `total_gvg_num` is better.

File: proto/base/types/gfsptask/task.proto [hunk 4]
```diff
@@ -201,6 +201,12 @@ message GfSpMigratePieceTask {
...
+  uint64 gvg_num_gc_finished = 7; // gc has finished gvg num
```
Comment (sysvm) [inline proto/base/types/gfsptask/task.proto:209]: `gc_finished_gvg_num`?

**Review by sysvm — APPROVED** [review]
LGTM

File: store/sqldb/migrate_schema.go [hunk 5]
```diff
@@ -38,9 +38,33 @@ type MigrateGVGTable struct {
...
 // TableName is used to set MigrateGVGTable Schema's table name in database.
 func (MigrateGVGTable) TableName() string {
 	return MigrateGVGTableName
 }
+
+// MigrateBucketProgressTable table schema.
+// used by persist bucket migration progress and meta
+type MigrateBucketProgressTable struct {
+	BucketID              uint64 `gorm:"primary_key"`
+	SubscribedBlockHeight uint64 `gorm:"primary_key"`
```
Comment (will-2012) [inline store/sqldb/migrate_schema.go:53]: What is the meaning of the two primary keys in gorm?

File: base/gfspclient/manager.go [hunk 6]
```diff
@@ -256,3 +256,20 @@ func (s *GfSpClient) ResetRecoveryFailedList(ctx context.Context) ([]string, err
 	}
 	return resp.GetRecoveryFailedList(), nil
 }
+
+func (s *GfSpClient) GetBucketMigrationState(ctx context.Context, bucketID uint64) (*gfspserver.MigrateBucketProgressMeta, error) {
```
Comment (will-2012) [inline base/gfspclient/manager.go:260]: state or progress, pls try to keep a consistent style.

File: core/spdb/entity.go [hunk 7]
```diff
@@ -119,8 +119,9 @@ type MigrateGVGUnitMeta struct {
 	SrcSPID                  uint32
 	DestSPID                 uint32
 	LastMigratedObjectID     uint64
-	MigrateStatus            int // scheduler assign unit status.
-	RetryTime                int //
+	MigrateStatus            int    // scheduler assign unit status.
```
Comment (will-2012) [inline core/spdb/entity.go:122]: status or state, pls distinguish semantics and be consistent.

File: modular/executor/execute_gc.go [hunk 8]
```diff
@@ -384,13 +384,16 @@ func (e *ExecuteModular) gcMetaReadRecord(ctx context.Context, task coretask.GCM
 }
 
 func (e *ExecuteModular) HandleGCBucketMigrationBucket(ctx context.Context, task coretask.GCBucketMigrationTask) {
```
Comment (will-2012) [inline modular/executor/execute_gc.go:386]: Whether to support continuing gc from the last breakpoint.

File: modular/gater/router.go [hunk 9]
```diff
@@ -169,6 +170,10 @@ func (g *GateModular) RegisterHandler(router *mux.Router) {
 		// Get Bucket Meta
 		r.NewRoute().Name(getBucketMetaRouterName).Methods(http.MethodGet).Queries(GetBucketMetaQuery, "").HandlerFunc(g.getBucketMetaHandler)
 
+		// Query migration progress
+		r.NewRoute().Name(queryMigrationProgressRouterName).Methods(http.MethodGet).HandlerFunc(g.queryBucketMigrationProgressHandler).
```
Comment (will-2012) [inline modular/gater/router.go:174]: Add ut.

File: store/sqldb/migrate.go [hunk 10]
```diff
@@ -418,3 +463,278 @@ func (s *SpDBImpl) DeleteMigrateGVGUnitsByBucketID(bucketID uint64) error {
...
+	return nil
+}
+
+func (s *SpDBImpl) QueryMigrateBucketState(bucketID uint64) (int, error) {
+	var (
+		result              *gorm.DB
+		insertMigrateBucket *MigrateBucketProgressTable
+		queryReturn         *MigrateBucketProgressTable
+		needInsert          bool
+		state               = 0
```
Comment (will-2012) [inline store/sqldb/migrate.go:509]: what's the meaning of 0.

File: store/sqldb/migrate.go [hunk 11]
```diff
@@ -418,3 +463,278 @@ func (s *SpDBImpl) DeleteMigrateGVGUnitsByBucketID(bucketID uint64) error {
...
+	)
+
+	progress := &spdb.MigrateBucketProgressMeta{}
+
+	queryReturn = &MigrateBucketProgressTable{}
+	result = s.db.First(queryReturn, "bucket_id = ?", bucketID)
+	if result.Error != nil && !errors.Is(result.Error, gorm.ErrRecordNotFound) {
+		return &spdb.MigrateBucketProgressMeta{}, result.Error
+	}
+	if errors.Is(result.Error, gorm.ErrRecordNotFound) {
```
Comment (will-2012) [inline store/sqldb/migrate.go:549]: What is the special meaning of this if? Which has been included in the revious if condition

File: store/sqldb/migrate.go [hunk 12]
```diff
@@ -418,3 +463,278 @@ func (s *SpDBImpl) DeleteMigrateGVGUnitsByBucketID(bucketID uint64) error {
...
+	progress.LastGCGvgID = queryReturn.LastGCGvgID
+	progress.RecoupQuota = queryReturn.RecoupQuota
+	progress.MigrationState = queryReturn.MigrationState
+
+	return progress, nil
+}
+
+func (s *SpDBImpl) ListBucketMigrationToConfirm() ([]*spdb.MigrateBucketProgressMeta, error) {
+	var queryReturns []MigrateBucketProgressTable
+	migrationStates := []int{11, 13}
```
Comment (will-2012) [inline store/sqldb/migrate.go:565]: use enum or const.

File: base/gfspapp/manage_server.go [hunk 13]
```diff
@@ -360,22 +360,48 @@ func (g *GfSpBaseApp) GfSpQueryTasksStats(ctx context.Context, _ *gfspserver.GfS
 	}, nil
 }
 
-func (g *GfSpBaseApp) GfSpNotifyPreMigrate(ctx context.Context, req *gfspserver.GfSpNotifyPreMigrateBucketRequest) (
+func (g *GfSpBaseApp) GfSpQueryBucketMigrationProgress(ctx context.Context, req *gfspserver.GfSpQueryBucketMigrationProgressRequest) (
```
Comment (ruojunm) [inline base/gfspapp/manage_server.go:363]: Do we have more testing in manage_server_test.go?

File: modular/gater/bucket_handler.go [hunk 14]
```diff
@@ -228,3 +230,91 @@ func (g *GateModular) listBucketReadRecordHandler(w http.ResponseWriter, r *http
 	}
 	log.Debugw("succeed to list bucket read records", "xml_info", xmlInfo)
 }
+
+// queryBucketMigrationProgressHandler handles the query bucket migration  progress request.
+func (g *GateModular) queryBucketMigrationProgressHandler(w http.ResponseWriter, r *http.Request) {
```
Comment (ruojunm) [inline modular/gater/bucket_handler.go:235]: please add UT in bucket_handler_test please

File: modular/gater/migrate_handler.go [hunk 15]
```diff
@@ -394,22 +396,19 @@ func (g *GateModular) getLatestBucketQuotaHandler(w http.ResponseWriter, r *http
...
-	bz, err := quota.Marshal()
-	if err != nil {
+	if bz, err = quota.Marshal(); err != nil {
 		log.CtxErrorw(reqCtx.Context(), "failed to marshal", "bucket_id",
 			bucketID, "error", err)
 		return
 	}
 
-	w.Header().Set(GnfdSignedApprovalMsgHeader, hex.EncodeToString(bz))
+	w.Header().Set(GnfdQuotaInfoHeader, hex.EncodeToString(bz))
```
Comment (ruojunm) [inline modular/gater/migrate_handler.go:411]: can we verify those changes in UT?

File: modular/metadata/metadata_sp_exit_service.go [hunk 16]
```diff
@@ -244,6 +244,55 @@ func (r *MetadataModular) GfSpListMigrateBucketEvents(ctx context.Context, req *
 	return resp, nil
 }
 
+// GfSpListCompleteMigrationBucketEvents list migrate bucket events, sp_id should be src sp id
+func (r *MetadataModular) GfSpListCompleteMigrationBucketEvents(ctx context.Context, req *types.GfSpListCompleteMigrationBucketEventsRequest) (resp *types.GfSpListCompleteMigrationBucketEventsResponse, err error) {
+	var (
```
Comment (ruojunm) [inline modular/metadata/metadata_sp_exit_service.go:249]: UT please

**Review by ruojunm — APPROVED** [review]

File: modular/executor/execute_gc.go [hunk 17]
```diff
@@ -400,51 +400,77 @@ func (e *ExecuteModular) HandleGCBucketMigrationBucket(ctx context.Context, task
...
-			return
-		}
-		// can delete, verify
-		for _, obj := range objects {
-			objectInfo := obj.GetObject().GetObjectInfo()
-			if e.gcWorker.checkGVGMatchSP(ctx, objectInfo, piecestore.PrimarySPRedundancyIndex) == ErrInvalidRedundancyIndex {
-				err = e.gcWorker.deleteObjectSegmentsAndIntegrity(ctx, objectInfo)
-				log.CtxInfow(ctx, "succeed to delete objects by gvg and bucket for gc", "object", objectInfo, "error", err)
+		for {
+			if objectList, err = e.baseApp.GfSpClient().ListObjectsByGVGAndBucketForGC(ctx, gvg.GetId(), bucketID, startAfter, limit); err != nil {
```
Comment (alexgao001) [inline modular/executor/execute_gc.go:439]: not find out `startAfter` is updated in this for loop.

File: modular/gater/migrate_handler.go [hunk 18]
```diff
@@ -451,44 +452,23 @@ func (g *GateModular) preMigrateBucketHandler(w http.ResponseWriter, r *http.Req
 	}
 
 	bucketID = bucketMigrationInfo.GetBucketId()
-	err = g.baseApp.GfSpClient().NotifyPreMigrateBucket(reqCtx.Context(), bucketID)
-	if err != nil {
+	if quota, err = g.baseApp.GfSpClient().NotifyPreMigrateBucketAndDeductQuota(reqCtx.Context(), bucketID); err != nil || quota == nil {
 		log.CtxErrorw(reqCtx.Context(), "failed to pre migrate bucket, the bucket may already notified", "bucket_id",
 			bucketID, "error", err)
+		// if the bucket has already pre notified ignore the error
+		if strings.Contains(err.Error(), "the bucket has already notified") {
```
Comment (alexgao001) [inline modular/gater/migrate_handler.go:459]: the err can be `nil` here?

File: modular/manager/bucket_migrate_scheduler.go [hunk 19]
```diff
@@ -526,32 +555,68 @@ func (s *BucketMigrateScheduler) doneMigrateBucket(bucketID uint64) error {
 	s.deleteExecutePlanByBucketID(bucketID)
 	executePlan.stopSPSchedule()
 	err = s.manager.baseApp.GfSpDB().DeleteMigrateGVGUnitsByBucketID(bucketID)
```
Comment (alexgao001) [inline modular/manager/bucket_migrate_scheduler.go:557]: would this records stay in DB if skipping?

File: modular/manager/bucket_migrate_scheduler.go [hunk 20]
```diff
@@ -560,35 +625,88 @@ func (s *BucketMigrateScheduler) processEvents(migrateBucketEvents *types.ListMi
...
+			return
+		}
+		if err = s.doneMigrateBucket(bucketID); err != nil {
+			log.Errorw("failed to done migrate bucket", "EventMigrationBucket", event, "error", err)
+			return
+		}
+		if err = UpdateBucketMigrationProgress(s.manager.baseApp, bucketID, storetypes.BucketMigrationState_MIGRATION_FINISHED); err != nil {
+			return
+		}
+		log.CtxInfow(ctx, "succeed to confirm complete events", "EventMigrationBucket", event, "error", err)
```
Comment (alexgao001) [inline modular/manager/bucket_migrate_scheduler.go:681]: no err

File: modular/manager/migrate_service.go [hunk 21]
```diff
@@ -45,44 +46,125 @@ func (m *ManageModular) NotifyMigrateSwapOut(ctx context.Context, swapOut *virtu
...
-			bucketID, "bucket_size", bucketSize, "error", err)
-		return err
+
+	// get bucket quota and check
+	if bucketSize, err = m.getBucketTotalSize(ctx, bucketID); err != nil {
+		log.Errorf("failed to get bucket total object size", "bucket_id", bucketID, "error", err)
+		return latestQuota, err
 	}
 
 	if bmInfo.GetFinished() {
```
Comment (alexgao001) [inline modular/manager/migrate_service.go:128]: seems this useless `if`

**Review by alexgao001 — APPROVED** [review]

File: base/gfspclient/gater.go [hunk 22]
```diff
@@ -243,7 +245,7 @@ func (s *GfSpClient) QueryLatestBucketQuota(ctx context.Context, endpoint string
 		return gfsptask.GfSpBucketQuotaInfo{}, fmt.Errorf("failed to query latest bucket quota, bucket(%s), status_code(%d), endpoint(%s)", queryMsg, resp.StatusCode, endpoint)
 	}
 
-	signedMsg, err := hex.DecodeString(resp.Header.Get(GnfdSignedApprovalMsgHeader))
+	signedMsg, err := hex.DecodeString(resp.Header.Get(GnfdQuotaInfoHeader))
```
Comment (will-2012) [inline base/gfspclient/gater.go:248]: `signedMsg` may be misleading, it can be optimized.

File: core/spdb/entity.go [hunk 23]
```diff
@@ -130,3 +131,20 @@ type SwapOutMeta struct {
 	SwapOutMsg    *virtualgrouptypes.MsgSwapOut
 	CompletedGVGs []uint32
 }
+
+// MigrateBucketProgressMeta is used to record migrate bucket progress meta.
+type MigrateBucketProgressMeta struct {
+	BucketID              uint64 // as primary key
+	SubscribedBlockHeight uint64
+	MigrationState        int
```
Comment (will-2012) [inline core/spdb/entity.go:139]: migrateState?

**Review by will-2012 — APPROVED** [review]

**Pull request merged**